src/Components/Web.JS/src/Boot.Web.ts

```diff
@@ -15,50 +15,113 @@ import { shouldAutoStart } from './BootCommon';
+import { RootComponentManager } from './Services/RootComponentManager';
+import { RendererId } from './Rendering/RendererId';
 
 let started = false;
+let isPerformingEnhnacedNavigation = false;
 let webStartOptions: Partial<WebStartOptions> | undefined;
+let lastActiveElement: Element | null = null;
 
-async function boot(options?: Partial<WebStartOptions>): Promise<void> {
+const circuitRootComponents = new RootComponentManager(RendererId.Server);
+const webAssemblyRootComponents = new RootComponentManager(RendererId.WebAssembly);
+
+function boot(options?: Partial<WebStartOptions>) : Promise<void> {
   if (started) {
     throw new Error('Blazor has already started.');
   }
   started = true;
   webStartOptions = options;
 
-  attachStreamingRenderingListener(options?.ssr);
+  const navigationEnhancementCallbacks: NavigationEnhancementCallbacks = {
+    beforeEnhancedNavigation,
+    beforeDocumentUpdated,
+    afterDocumentUpdated,
+    afterEnhancedNavigation,
+  };
+
+  const descriptorHandler: DescriptorHandler = {
+    onDescriptorAdded,
+  };
+
+  attachComponentDescriptorHandler(descriptorHandler);
+  attachStreamingRenderingListener(options?.ssr, navigationEnhancementCallbacks);
 
   if (!options?.ssr?.disableDomPreservation) {
-    attachProgressivelyEnhancedNavigationListener(activateInteractiveComponents);
+    attachProgressivelyEnhancedNavigationListener(navigationEnhancementCallbacks);
   }
 
-  await activateInteractiveComponents();
+  processComponentDescriptors(document);
+
+  return Promise.resolve();
+}
+
+function onDescriptorAdded(descriptor: ServerComponentDescriptor | WebAssemblyComponentDescriptor) {
+  switch (descriptor.type) {
+    case 'server':
+      // Start the circuit now so that it's more likely to be ready by the time interactivity actually starts.
+      startCircuitIfNotStarted();
+      circuitRootComponents.registerComponentDescriptor(descriptor);
+      break;
+    case 'webassembly':
+      startWebAssemblyIfNotStarted();
+      webAssemblyRootComponents.registerComponentDescriptor(descriptor);
+      break;
+  }
 }
 
-async function activateInteractiveComponents() {
-  const serverComponents = discoverComponents(document, 'server') as ServerComponentDescriptor[];
-  const webAssemblyComponents = discoverComponents(document, 'webassembly') as WebAssemblyComponentDescriptor[];
+function beforeEnhancedNavigation() {
+  isPerformingEnhnacedNavigation = true;
+}
 
-  if (serverComponents.length) {
-    // TEMPORARY until https://github.com/dotnet/aspnetcore/issues/48763 is implemented
-    // As soon we we see you have interactive components, we'll stop doing enhanced nav even if you don't have an interactive router
-    // This is because, otherwise, we would need a way to add new interactive root components to an existing circuit and that's #48763
-    detachProgressivelyEnhancedNavigationListener();
+function beforeDocumentUpdated() {
+  lastActiveElement = document.activeElement;
```

Comment:
Hopefully this won't be required once we start leaving the interactive-component DOM in the document while merging in the new static content.